### PR TITLE
[3.x][Installer] Add upload progress and improve error handling

### PR DESCRIPTION
**Pull request by asika32764**
Add upload progress so user can see the upload percentage like before.

p-2017-11-12-002

Improve error handler so user will get message if PHP raise fatal error without standard JSON format or HTTP error status.

p-2017-11-12-003

### CODE REVIEW

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 1]
```diff
@@ -130,10 +155,30 @@
...
+				} else {
+					showError(error.statusText);
+				}
 			});
+
+			function showError(res) {
+				actions.show();
+				progress.hide();
+
+				var message = 'Unknown error.';
```
Comment (brianteeman) [inline plugins/installer/packageinstaller/tmpl/default.php]: shouldnt this  be translatable?

**Comment by brianteeman** [comment]
It would be nice if this was accessible  there are lots of accessible progress bar examples online

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 2]
```diff
@@ -130,10 +155,30 @@
...
+
+				if (typeof res === 'string') {
+					// Let's remove unnecessary HTML
+					message = res.replace(/(<([^>]+)>|\s+)/g, ' ');
+				} else if (res.message) {
+					message = res.message;
+				}
+
+				alert(message);
```
Comment (Fedik) [inline plugins/installer/packageinstaller/tmpl/default.php]: maybe would be better `Joomla.renderMessages()`?

File: plugins/installer/packageinstaller/tmpl/default.php (hunk 1 above)
Comment (asika32764) [reply plugins/installer/packageinstaller/tmpl/default.php]: Hmm, this is only a fallback default message, should never display. (If this message display, it means an extension is on developing and developer try to make it error with empty or non-JSON output...) 

Do you think it is a good idea to translate a message which may only developers can see? If you think so I can translate it.

**Comment by asika32764** [comment]
@brianteeman Add accessible.

@Fedik Use Joomla.renderMessages()

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 3]
```diff
@@ -61,6 +66,10 @@
 		});
 		
 		fileInput.on('change', function (e) {
+			if (uploading) {
+				return;
+			}
+
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Remove tabs in line 72.

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 4]
```diff
@@ -96,6 +105,10 @@
 			e.stopPropagation();
 
 			dragZone.removeClass('hover');
+
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Remove tabs in line 108.

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 5]
```diff
@@ -108,20 +121,44 @@
...
 				processData: false,
 				cache: false,
-				contentType: false
-			}).done(function (res) {
-				if (!res.success) {
-					console.log(res.message, res.messages);
+				contentType: false,
+				xhr: function () {
+					var xhr = new window.XMLHttpRequest();
+
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Remove tabs in line 138.

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 6]
```diff
@@ -108,20 +121,44 @@
...
+					xhr.upload.addEventListener("progress", function (evt) {
+						if (evt.lengthComputable) {
+							var percentComplete = evt.loaded / evt.total;
+							var number = Math.round(percentComplete * 100);
+							progressBar.css('width', number + '%');
+							progressBar.attr('aria-valuenow', number);
+							percentage.text(number);
+						}
+					}, false);
+
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Remove tabs in line 153.

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on b82544e892ba9cb0f2262b33c3f6105499003438<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

**Comment by ghost** [comment]
@asika32764 how to get "fatal error without standard JSON format or HTTP error status" to test PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

**Comment by asika32764** [comment]
Actually I don't know. Maybe return a XML output. This case should never happened.

In this PR, all response with HTTP 200 will go to `done()` method, but the PHP post_max_size error is unable to catch and returned with JSON and HTTP 200 status:

p-2017-11-12-005

If jQuery try to parse JSON but fail, it will go to `error()` method. And if response is HTTP 4xx-5xx status will go `error()` too. I handled the `error.responseText` for any PHP fatal error and `error.responseJSON` for standard JSON output, so it should has no other cases we will get.

**Comment by asika32764** [comment]
@franz-wohlkoenig Sorry, I misunderstand what you asked. The  "fatal error without standard JSON format or HTTP error status" is the `post_max_size` issue.

Set php.ini post_max_size to a small value and upload a big file, you will get error message. Before this PR, the uploader will return no message.

If you upload an extension with fatal error in it's install script, this PR will also show the error message. Before this PR it will lose respond and running forever.

**Comment by Quy** [comment]
This works with drag and drop, but not with `Or browse for file`. Is this intentional?

**Comment by asika32764** [comment]
Yes, `Or browse for file` use file input and post form to upload package. While the drag and drop use ajax (because JS cannot insert value to file input).

File: plugins/installer/packageinstaller/tmpl/default.php (hunk 1 above)
Comment (brianteeman) [reply plugins/installer/packageinstaller/tmpl/default.php]: Personally yes i do

Comment (asika32764) [reply plugins/installer/packageinstaller/tmpl/default.php]: OK let's translate it.

**Comment by Quy** [comment]
On a local install, the progress bar moves to 100% very quickly, and then takes about 10 seconds to complete. So this is an upload indicator only and not an installing indicator too?

**Comment by asika32764** [comment]
Yes, it is only uploading indicator. Joomla didn't implement the installing indicator interface now.

**Comment by Quy** [comment]
In Microsoft Edge 41.16299.15.0 on Windows 10, you don't see the progress bar until the very end. Also, should the % be next to the number without the space?

edge-upload

**Comment by brianteeman** [comment]
good spot. yes it would read better without the space between the number and the &

**Comment by asika32764** [comment]
- Removed space
- Added installing progress after uploaded. See:

up-pro

---- 

@Quy About the Edge issue, it is a known bug that MS haven't fixed it. See:

- https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/12224510/
- https://stackoverflow.com/questions/43942396/ajax-file-upload-progress-upload-event-does-not-fire-in-edge-browser

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 7]
```diff
@@ -108,20 +123,47 @@
...
+					// Upload progress
+					xhr.upload.addEventListener("progress", function (evt) {
+						if (evt.lengthComputable) {
+							var percentComplete = evt.loaded / evt.total;
+							var number = Math.round(percentComplete * 100);
+							progressBar.css('width', number + '%');
+							progressBar.attr('aria-valuenow', number);
+							percentage.text(number);
+
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Remove tabs on line 152.

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 8]
```diff
@@ -181,23 +283,52 @@
...
-			</p>
-			<p>
-				<?php echo JText::sprintf('JGLOBAL_MAXIMUM_UPLOAD_SIZE_LIMIT', $maxSize); ?>
-			</p>
+			<div class="upload-progress">
+				<div class="progress progress-striped active">
+					<div class="bar bar-success"
+						style="width: 0;"
+						role="progressbar"
+						aria-valuenow="25"
```
Comment (Quy) [inline plugins/installer/packageinstaller/tmpl/default.php]: Should the value be 0?

Comment (asika32764) [reply plugins/installer/packageinstaller/tmpl/default.php]: The `width` value? The initial width value should be `0`.

Comment (asika32764) [reply plugins/installer/packageinstaller/tmpl/default.php]: Ah, I know, you mean the `aria-valuenow`

Yes it should be 0. Fixed.

File: administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini [hunk 9]
```diff
@@ -5,9 +5,13 @@
 
 PLG_INSTALLER_PACKAGEINSTALLER_DRAG_FILE_HERE="Drag and drop file here to upload."
 PLG_INSTALLER_PACKAGEINSTALLER_EXTENSION_PACKAGE_FILE="Extension package file"
+PLG_INSTALLER_PACKAGEINSTALLER_INSTALLING="Installing..."
 PLG_INSTALLER_PACKAGEINSTALLER_NO_PACKAGE="Please select a package to upload"
 PLG_INSTALLER_PACKAGEINSTALLER_PLUGIN_XML_DESCRIPTION="This plugin allows you to install packages from your local computer."
 PLG_INSTALLER_PACKAGEINSTALLER_SELECT_FILE="Or browse for file"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini:12]: Sorry to nitpick. Please move lines 14-16 below line 11.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on a00243c9dba19d6c047c2d24b83b22045d07d739<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

File: administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini [hunk 10]
```diff
@@ -5,9 +5,13 @@
...
+PLG_INSTALLER_PACKAGEINSTALLER_INSTALLING="Installing..."
 PLG_INSTALLER_PACKAGEINSTALLER_NO_PACKAGE="Please select a package to upload"
 PLG_INSTALLER_PACKAGEINSTALLER_PLUGIN_XML_DESCRIPTION="This plugin allows you to install packages from your local computer."
 PLG_INSTALLER_PACKAGEINSTALLER_SELECT_FILE="Or browse for file"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_EMPTY="Error: Server returns empty response."
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_UNKNOWN="Error: Unknown error or invalid JSON output."
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
-PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini]: For silly reasons it is our practice not to remove unused strings. So please put this string back. You can if you want to add a comment above it
;The next line is deprecated and will be removed in J4

**Review by brianteeman — CHANGES_REQUESTED** [review]

File: administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini (hunk 10 above)
Comment (asika32764) [reply administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini]: OK added the deprecated info

**Review by brianteeman — APPROVED** [review]

File: administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini [hunk 11]
```diff
@@ -5,9 +5,16 @@
 
 PLG_INSTALLER_PACKAGEINSTALLER_DRAG_FILE_HERE="Drag and drop file here to upload."
 PLG_INSTALLER_PACKAGEINSTALLER_EXTENSION_PACKAGE_FILE="Extension package file"
+PLG_INSTALLER_PACKAGEINSTALLER_INSTALLING="Installing..."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini]: Please add space before `...`

File: administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini [hunk 12]
```diff
@@ -5,9 +5,16 @@
...
 PLG_INSTALLER_PACKAGEINSTALLER_NO_PACKAGE="Please select a package to upload"
 PLG_INSTALLER_PACKAGEINSTALLER_PLUGIN_XML_DESCRIPTION="This plugin allows you to install packages from your local computer."
 PLG_INSTALLER_PACKAGEINSTALLER_SELECT_FILE="Or browse for file"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_AND_INSTALL="Upload & Install"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_EMPTY="Error: Server returns empty response."
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_ERROR_UNKNOWN="Error: Unknown error or invalid JSON output."
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_INSTALL_JOOMLA_EXTENSION="Upload & Install Joomla Extension"
 PLG_INSTALLER_PACKAGEINSTALLER_UPLOAD_PACKAGE_FILE="Upload Package File"
+PLG_INSTALLER_PACKAGEINSTALLER_UPLOADING="Uploading..."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_installer_packageinstaller.ini]: Please add space before `...`

**Comment by asika32764** [comment]
ok

**Review by mbabker — APPROVED** [review]

**Comment by twister65** [comment]
I have tested this item :red_circle: unsuccessfully on 7b6ed585486c7da22f2b7c0d8930477bd94baff8<br /><br />I do not have an alert popup when the package size exceeds the upload_max_filesize value.
Nothing happening, I have only this message : "No Install Type Found".
This problem is recurring on my host where the upload_max_filesize is set to 2Mo.
Great PR, but it does not work as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

**Comment by asika32764** [comment]
@twister65

The test case with error message is `post_max_size`.

PHP won't catch `upload_max_size` error, it will still run process without any uploaded file, so Joomla cannot find extension package file and return a `No Install Type Found` message.

See https://github.com/joomla/joomla-cms/pull/18554#issuecomment-343742342

**Comment by asika32764** [comment]
At here we use message instead of alert popup.
https://github.com/joomla/joomla-cms/pull/18554#issuecomment-343708464

**Comment by twister65** [comment]
@asika32764 , if I set post_max_size to 2 MB, and I upload a file larger than 2 MB. Then the message is "No Install Type Found" and there is no alert popup. This message is not relevant.

**Comment by asika32764** [comment]
> Then the message is "No Install Type Found" and there is no alert popup. 

This is expected behavior,the default error message should show as Joomla message, not popup.

This PR try to catch some PHP error which Joomla cannot catch and notify user, without this PR, the uploader will only infinite loop without any response. 

The popup screenshot you see in first post is no longer used, we use message box now. See https://github.com/joomla/joomla-cms/pull/18554#issuecomment-343708464

> This message is not relevant.

This message is default Joomla install error message, not belongs to this PR.

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 7b6ed585486c7da22f2b7c0d8930477bd94baff8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

File: plugins/installer/packageinstaller/tmpl/default.php [hunk 13]
```diff
@@ -181,23 +283,52 @@
...
+					<button id="select-file-button" type="button" class="btn btn-success">
+						<span class="icon-copy" aria-hidden="true"></span>
+						<?php echo JText::_('PLG_INSTALLER_PACKAGEINSTALLER_SELECT_FILE'); ?>
+					</button>
+				</p>
+				<p>
+					<?php echo JText::sprintf('JGLOBAL_MAXIMUM_UPLOAD_SIZE_LIMIT', $maxSize); ?>
+				</p>
+			</div>
 		</div>
```
Comment (infograf768) [inline plugins/installer/packageinstaller/tmpl/default.php]: Maybe delete empty line

Comment (asika32764) [reply plugins/installer/packageinstaller/tmpl/default.php]: Deleted

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 5a68c03c94069f0189d598d5007ad86504a0fd97<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/18554">issues.joomla.org/tracker/joomla-cms/18554</a>.</sub>

**Comment by infograf768** [comment]
As the last change is just cosmetic, let's keep this as RTC.

**Pull request merged**